lightning/src/ln/channelmanager.rs

```diff
@@ -8831,32 +8831,47 @@ where
 			.map_err(|_| Bolt12SemanticError::DuplicatePaymentId)?;
 
 		let mut pending_offers_messages = self.pending_offers_messages.lock().unwrap();
-		if !offer.paths().is_empty() {
-			// Send as many invoice requests as there are paths in the offer (with an upper bound).
-			// Using only one path could result in a failure if the path no longer exists. But only
-			// one invoice for a given payment id will be paid, even if more than one is received.
-			const REQUEST_LIMIT: usize = 10;
-			for path in offer.paths().into_iter().take(REQUEST_LIMIT) {
-				let message = new_pending_onion_message(
-					OffersMessage::InvoiceRequest(invoice_request.clone()),
-					Destination::BlindedPath(path.clone()),
-					Some(reply_path.clone()),
-				);
-				pending_offers_messages.push(message);
-			}
-		} else if let Some(signing_pubkey) = offer.signing_pubkey() {
-			let message = new_pending_onion_message(
-				OffersMessage::InvoiceRequest(invoice_request),
-				Destination::Node(signing_pubkey),
-				Some(reply_path),
-			);
-			pending_offers_messages.push(message);
-		} else {
+		pending_offers_messages.extend(self.create_invoice_request_messages(invoice_request, reply_path)?);
+
+		Ok(())
+	}
+
+	fn create_invoice_request_messages(
+		&self, invoice_request: InvoiceRequest, reply_path: BlindedPath
+	) -> Result<Vec<PendingOnionMessage<OffersMessage>>, Bolt12SemanticError> {
+		let paths = invoice_request.paths();
+		let signing_pubkey = invoice_request.signing_pubkey();
+
+		if paths.is_empty() && signing_pubkey.is_none() {
 			debug_assert!(false);
 			return Err(Bolt12SemanticError::MissingSigningPubkey);
 		}
 
-		Ok(())
+		// Send as many invoice requests as there are paths in the offer (with an upper bound).
+		// Using only one path could result in a failure if the path no longer exists. But only
+		// one invoice for a given payment id will be paid, even if more than one is received.
+		const REQUEST_LIMIT: usize = 10;
+
+		let messages = if !paths.is_empty() {
+			paths.into_iter()
+				.take(REQUEST_LIMIT)
+				.map(|path| {
+					new_pending_onion_message(
+						OffersMessage::InvoiceRequest(invoice_request.clone()),
+						Destination::BlindedPath(path.clone()),
+						Some(reply_path.clone()),
+					)
+				})
+				.collect()
+		} else {
+			vec![new_pending_onion_message(
+				OffersMessage::InvoiceRequest(invoice_request.clone()),
+				Destination::Node(signing_pubkey.unwrap()),
+				Some(reply_path.clone()),
+			)]
+		};
```

Comment:
Instead of allocating vectors, could we simply `push` into `pending_offers_messages`? Could call the method `enqueue_invoice_requests`.

Reply:
Makes sense!
Updated in [pr3010.19](https://github.com/shaavan/rust-lightning/commits/pr3010.19) to directly push to pending_offers_messages instead of allocating a vector. Also renamed the function to `enqueue_invoice_requests`
Thanks for the suggestion!